File health: trailing newlines, whitespace, etc fixes and GitHub Actions workflow

FranciscoPombal opened the pull request:
Continuation of https://github.com/qbittorrent/qBittorrent/pull/13236 with the ideas discussed there and without the translations bit. The translations stuff can be handled later (@sledgehammer999, if you please); https://github.com/qbittorrent/qBittorrent/issues/12369 is waiting on translations to get fixed.

---

What this does:

- The first 3 commits are just manual "file health" fixes
- The last commit introduces a new workflow that checks for "file health" regressions on any PR or commit to any branch, so that they don't happen again and don't accumulate from here on out. Thus, this matter is settled once and for all.

FranciscoPombal commented:
**Note:** The new workflow is currently expected to fail due to some offending translation files (since no translation file was touched at all). Here is the result: https://github.com/qbittorrent/qBittorrent/pull/13659/checks?check_run_id=1310098147

glassez commented on .github/workflows/file_health_exclusions_nonutf8.txt:
```diff
@@ -0,0 +1,85 @@
+./dist/mac/CMakeLists.txt
```
Why is it non-UTF8?

FranciscoPombal replied:
@glassez 

Because it's empty. With the current method (`file --mime`), empty files can't be considered UTF-8, regardless of their purpose or extension:

```
$ file --mime dist/mac/CMakeLists.txt 
dist/mac/CMakeLists.txt: inode/x-empty; charset=binary
```

FranciscoPombal replied:
fixed.

glassez commented on .github/workflows/file_health_exclusions_nonutf8.txt:
```diff
@@ -0,0 +1,85 @@
+./dist/windows/nsis plugins/FindProcDLL_mod_by_hnedka.7z
+./dist/windows/nsis plugins/nsisFirewall.zip
+./dist/windows/nsis plugins/UAC.zip
+./.github/workflows/file_health_exclusions_cr.txt
```
Why are all these text files non-UTF8?

FranciscoPombal replied:
@glassez 

Same reasoning as https://github.com/qbittorrent/qBittorrent/pull/13659#discussion_r512657718 - because they're empty. If/when they stop being just empty files, we'll remove them from the exclusions.

glassez replied:
>If/when they stop being just empty files, we'll remove them from the exclusions.

Not so good reasoning... It can be simply forgotten.
>Because it's empty. With the current method (`file --mime`), empty files can't be considered UTF-8, regardless of their purpose or extension

Is it hard to check for file size before?

FranciscoPombal replied:
fixed.

glassez commented on .github/workflows/file_health_exclusions_nonutf8.txt:
```diff
@@ -0,0 +1,85 @@
+./dist/mac/CMakeLists.txt
+./dist/mac/qBitTorrentDocument.icns
+./dist/mac/qbittorrent_mac.icns
+./dist/qt-translations/qtbase_ar.qm
```
It would be better to use patterns here.

glassez replied:
Or even hardcode exclusions of some file types.

FranciscoPombal replied:
@glassez 

I'm afraid that could complicate things significantly for now, or at least it would force me to think about a different implementation. I think that's a good suggestion for a future improvement though. Let's keep things simple for now?

glassez replied:
It seems unreasonable to me that you should list every file if all files of this type should be excluded. This looks bad even for the initial implementation.

FranciscoPombal replied:
fixed.

FranciscoPombal commented:
@glassez 

Addressed all points. Previous solution was somewhat over-engineered.

- Simplified the checks
- Got rid of the exclusions files, now the exclusions are simply grep extended regular expressions in some variables in the script. This saves a ton of time and is always maintainable as long as one has at least heard of regular expressions.
- Made sure no files are empty (this is a new separate commit).

Chocobo1 commented on CODING_GUIDELINES.md:
```diff
@@ -1,10 +1,10 @@
-﻿All new code **must** follow the following coding guidelines.  
```
Trimming those whitespaces changed the result layout. 
Before: https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md
After: https://github.com/qbittorrent/qBittorrent/blob/e1fa7520721880447884e2d3f8591ed732a2bd61/CODING_GUIDELINES.md
You should replace them with a trailing `\` instead.

FranciscoPombal replied:
fixed.

Chocobo1 commented on .github/workflows/file_health.sh:
```diff
@@ -0,0 +1,49 @@
+
+echo -e "*** Detect files not encoded in UTF-8 ***\n"
+
+find . -path ./build -prune -false -o -path ./.git -prune -false -o -type f -exec file --mime {} \; | grep -E -v -e "charset=us-ascii" -e "charset=utf-8" -e "${exclusions_nonutf8}" | cut -d ":" -f 1 | sort | tee >(echo -e "\n--> Files not encoded in UTF-8: found" $(wc -l </dev/stdin) "regression(s)") | xargs -I my_input -0 bash -c 'echo "my_input"; test "$(echo -n "my_input" | wc -l)" -eq 0'; reg ...
```
Can you split them into mulitlines?
For example:
```bash
find . -path ./build -prune -false -o -path ./.git -prune -false -o -type f -exec file --mime {} \; \
    | grep -E -v -e "charset=us-ascii" -e "charset=utf-8" -e "${exclusions_nonutf8}" \
    | cut -d ":" -f 1 
```

FranciscoPombal replied:
fixed.

FranciscoPombal commented:
@glassez @Chocobo1 

Comments addressed.

https://github.com/qbittorrent/qBittorrent/pull/13659/commits/244fdb342f79be65c7ca6edf1d980f7785b88b81 now also contains a minor facelift/formatting fix to `CODING_GUIDELINES.md`. Basically, section headings were fixed, missing/extra newlines between constructs was fixed (and all `<br/>` removed), indentation of list elements was fixed, and ` \` was added at the end of some lines to keep the previous paragraph structure, even without trailing whitespace. The file now passes both the file health script and `markdownlint` without warnings.

The file health script also passes `shellcheck` without warnings, except for [SC2016](https://github.com/koalaman/shellcheck/wiki/SC2016), which can safely be considered a false-positive due to the `bash -c` invocations.

Chocobo1 reviewed (commented):
Also the check is still failing in this PR. You should exclude *.ts files, those are not for manual editing (at least in the current workflow).

FranciscoPombal commented:
@Chocobo1 

> Also the check is still failing in this PR. You should exclude *.ts files, those are not for manual editing (at least in the current workflow).

As a reminder, the problem of badly-formatted *`.ts` files, which has a ridiculously simple solution (format them properly), makes the program completely unusable for some users: https://github.com/qbittorrent/qBittorrent/issues/12369. @sledgehammer999 please address this, as mentioned in the OP. We need an urgent manual fix to the files, and then follow up this PR to add a check for `^</translation>`.

I left the whitespace warnings for `*.ts` files on for this PR (with the intent that this is merged with such warnings turned on, https://github.com/qbittorrent/qBittorrent/pull/13659#issuecomment-716681662) as an attempt to sound the alarm on this issue. Maybe a big red :x: on each CI run will finally annoy everyone enough to get this issue properly fixed.

FranciscoPombal commented on .github/workflows/file_health.sh:
```diff
@@ -0,0 +1,76 @@
+find . -path ./build -prune -false -o -path ./.git -prune -false -o -type f -exec file --mime {} \; | sort \
+    | grep -v -e "charset=us-ascii" -e "charset=utf-8" | cut -d ":" -f 1 \
+    | grep -E -v -e "${exclusions_nonutf8}" \
+    | tee >(echo -e "\n--> Files not encoded in UTF-8: found" "$(wc -l </dev/stdin)" "regression(s)") \
```
fixed. Though the `>` is not redirection here, it's process substitution (i.e. `tee >(...`), which does not allow a space between the angle bracket and the parentheses.

FranciscoPombal commented on .github/workflows/file_health.sh:
```diff
@@ -0,0 +1,76 @@
+    | grep -v -e "charset=us-ascii" -e "charset=utf-8" | cut -d ":" -f 1 \
+    | grep -E -v -e "${exclusions_nonutf8}" \
+    | tee >(echo -e "\n--> Files not encoded in UTF-8: found" "$(wc -l </dev/stdin)" "regression(s)") \
+    | xargs -I my_input -0 bash -c 'echo "my_input"; test "$(echo -n "my_input" | wc -l)" -eq 0';
```
fixed.

FranciscoPombal commented on CODING_GUIDELINES.md:
```diff
@@ -150,10 +161,12 @@ while (p) {
+
 When the conditional statement in `if`/`else` has only one line and its body occupy only one line,
-this also applies to loops statements.  
+this also applies to loops statements.
```
fixed.

FranciscoPombal commented on CODING_GUIDELINES.md:
```diff
@@ -338,83 +370,82 @@ class ExampleWidget : public QWidget
-a = (b <= MAX_B ? b : MAX_B);
-++a;
---b;
+  * container iteration and casts:
```
fixed.

FranciscoPombal commented on .github/workflows/file_health.sh:
```diff
@@ -0,0 +1,76 @@
+exclusions_tw='src/webui/www/private/scripts/lib/mootools-1.2-more.js'
+exclusions_no_lf='(compile_commands.json)|(src/webui/www/private/scripts/lib/mootools-1.2-(core-yc.js|more.js))|(.*svg)'
+
+echo -e "*** Detect files not encoded in UTF-8 ***\n"
```
@Chocobo1 meh, shell variables are global by default, unless you declare them as `local`. I don't think there is a big improvement in wrapping these simple one-liners in functions. Plus, I think for such a simple script, it is rather convenient to have all the exclusion regexes defined at the top (they are what will need to changed most often).

I think a V2 of this script could improve on the current implementation by reducing the repetition in each pipeline using functions, though (i.e., encapsulating some of the common pipeline logic, such as the final `| tee >(echo -e "\n--> Foo: found" "$(wc -l < /dev/stdin)" "regression(s)") \
    | xargs -I my_input -0 bash -c 'echo "my_input"; test "$(echo -n "my_input" | wc -l)" -eq 0'`)`.

Chocobo1 commented:
>I left the whitespace warnings for *.ts files on for this PR (with the intent that this is merged with such warnings turned on, #13659 (comment)) as an attempt to sound the alarm on this issue. Maybe a big red ❌ on each CI run will finally annoy everyone enough to get this issue properly fixed.

I would propose that we reintroduce check for *.ts files in a later PR so we can guard against other pending PRs first and defer fixing the root cause. It is unlikely for me to approve a PR with currently failing checks.

Also can you print out which line is failing for each checked file? Currently the checker only shows the file name but we need to be sure which line it is complaining.

FranciscoPombal commented:
@Chocobo1 

> I would propose that we reintroduce check for *.ts files in a later PR so we can guard against other pending PRs first and defer fixing the root cause. It is unlikely for me to approve a PR with currently failing checks.

Alright, fixed. But we really need to fix the translations problem.

Also, I noticed the current implementation only works correctly on zsh, and without `set -o pipefail`. Previously, the script would always fail, even if there were no regressions. This is now fixed. As for why it doesn't work in bash even without `set -o pipefail`, the reason seems to be tied to this: https://unix.stackexchange.com/a/406343. The zsh approach looks much clearer to me, and at this stage I'd rather not rethink the whole pipeline and syntax just to be able to use bash instead of zsh. Of course, it is slightly better to depend on a lower common denominator such as bash, but zsh isn't going anywhere anytime soon either.

> Also can you print out which line is failing for each checked file? Currently the checker only shows the file name but we need to be sure which line it is complaining.

I don't think that is relevant for these checks though, they don't require manual intervention, whether there is 1 problematic line or 100. For all of these tests, any decent text editor either fixes these automagically or has a "one-click" action that fixes the problem; alternatively, there is a standard Unix utility (`sed`, `awk`, etc) one-liner that can do the same:

- UTF-8 encoding problems -> "Save file with UTF-8 encoding" or similar, `iconv`, etc.
- Presence of `CR` byte (meaning lines ending with `CRLF`) -> "Change end of line sequence" or similar, `sed`, `awk`, etc.
- Trailing whitespace/too many trailing newlines/no trailing newline -> Any decent editor has an option to auto fix all of these on save, and of course it is also possible to fix wholesale with standard Unix utilities.

Chocobo1 commented:
@FranciscoPombal 
Thank you!
